src/metabase/task/sync_databases.clj

```diff
@@ -238,63 +241,71 @@
 (def ^:private sync-analyze-job (job sync-analyze-task-info))
 (def ^:private field-values-job (job field-values-task-info))
 
-(mu/defn ^:private trigger :- (ms/InstanceOfClass CronTrigger)
-  "Build a Quartz Trigger for `database` and `task-info`."
+(mu/defn ^:private trigger :- [:maybe (ms/InstanceOfClass CronTrigger)]
+  "Build a Quartz Trigger for `database` and `task-info` if a schedule exists."
   ^CronTrigger [database  :- (ms/InstanceOf Database)
                 task-info :- TaskInfo]
-  (triggers/build
-   (triggers/with-description (trigger-description database task-info))
-   (triggers/with-identity (trigger-key database task-info))
-   (triggers/using-job-data {"db-id" (u/the-id database)})
-   (triggers/for-job (job-key task-info))
-   (triggers/start-now)
-   (triggers/with-schedule
-     (cron/schedule
-      (cron/cron-schedule (cron-schedule database task-info))
-      ;; if we miss a sync for one reason or another (such as system being down) do not try to run the sync again.
-      ;; Just wait until the next sync cycle.
-      ;;
-      ;; See https://www.nurkiewicz.com/2012/04/quartz-scheduler-misfire-instructions.html for more info
-      (cron/with-misfire-handling-instruction-do-nothing)))))
+  (when-let [task-schedule (cron-schedule database task-info)]
+    (triggers/build
+     (triggers/with-description (trigger-description database task-info))
+     (triggers/with-identity (trigger-key database task-info))
+     (triggers/using-job-data {"db-id" (u/the-id database)})
+     (triggers/for-job (job-key task-info))
+     (triggers/start-now)
+     (triggers/with-schedule
+       (cron/schedule
+        (cron/cron-schedule task-schedule)
+        ;; if we miss a sync for one reason or another (such as system being down) do not try to run the sync again.
+        ;; Just wait until the next sync cycle.
+        ;;
+        ;; See https://www.nurkiewicz.com/2012/04/quartz-scheduler-misfire-instructions.html for more info
+        (cron/with-misfire-handling-instruction-do-nothing))))))
+
+(defn- update-db-trigger-if-needed!
+  "Replace or remove the existing trigger if the schedule changes, do nothing if schedule is the same."
+  [database task-info]
+  (let [job                                 (task/job-info (job-key task-info))
+        new-trigger                         (trigger database task-info)
+        ;; there should be only one schedule per task per DB
+        existing-trigger-with-same-schedule (when new-trigger
+                                              (some #(when (and (= (:key %) (.. new-trigger getKey getName))
```

Comment:
nitpick: I think this would be better as a `(first (filter ...))`, so you can avoid the awkward `when` block. 

I'd also hoist out `trigger-key` (`(.. new-trigger getKey getName)`) and `task-schedule` (`(cron-schedule database task-info)`) since those don't change, and they're making the predicate quite dense.

Reply:
the `when` part is needed else `(.. new-trigger .getKey)` will throw an error because `new-trigger` is nil.

Agree with hosting out `triger-key` and `task-schedule`. I'll make a nested `let`.

Reply:
You could also just make trigger-key nullable by using a when in its definition, if you want to save nesting.